## Условие 1: 1 задача

###### Скачать данные по ссылке https://www.kaggle.com/datasets/shivachandel/kc-house-data

In [10]:
import pandas as pd

###### Считать данные с помощью pandas

In [11]:
houses = pd.read_csv('./kc_house_data.csv', encoding='UTF-8')

###### Вывести на экран первые 5 строк

In [12]:
houses.head(5)

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

###### Посмотреть на описание признаков и на их содержание

In [13]:
houses.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

'id' - ид.номер  
'date' - дата покупки  
'price' - цена   
'bedrooms' - кол-во спальных комнат  
'bathrooms' - кол-во ванных комнат  
'sqft_living' - жилая площадь  
'sqft_lot' - общая площадь   
'floors' - кол-во этажей   
'waterfront' - наличие вида на море(воду)   
'view' - кол-во окон смотряших в разные стороны   
'condition' - состояние дома (1-5)  
'grade' - оценка (1-13)  
'sqft_above' - площадь крыши ??   
'sqft_basement' - площадь подвала   
'yr_built' - год постройки   
'yr_renovated' - год ремонта   
'zipcode' - почтовый индекс  
'lat' - широта  
'long' - долгота   
'sqft_living15' - ???  
'sqft_lot15' - ???  


## Условие 2: 2 задача

###### Проведите первичный анализ данных

In [14]:
houses.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21611.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.396095     291.509045   1971.005136     84.402258   
std        1.175459    828.128162     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

###### Изучите типы данных

In [15]:
houses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
id               21613 non-null int64
date             21613 non-null object
price            21613 non-null float64
bedrooms         21613 non-null int64
bathrooms        21613 non-null float64
sqft_living      21613 non-null int64
sqft_lot         21613 non-null int64
floors           21613 non-null float64
waterfront       21613 non-null int64
view             21613 non-null int64
condition        21613 non-null int64
grade            21613 non-null int64
sqft_above       21611 non-null float64
sqft_basement    21613 non-null int64
yr_built         21613 non-null int64
yr_renovated     21613 non-null int64
zipcode          21613 non-null int64
lat              21613 non-null float64
long             21613 non-null float64
sqft_living15    21613 non-null int64
sqft_lot15       21613 non-null int64
dtypes: float64(6), int64(14), object(1)
memory usage: 3.5+ MB


###### Найдите количество пропущенных ячеек в данных

Есть 2 пропущенных значения в столбце 'sqft_above'

###### Посчитайте основные статистики по всем признакам и поизучайте их

In [17]:
houses[['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated']].mean()

id               4.580302e+09
price            5.400881e+05
bedrooms         3.370842e+00
bathrooms        2.114757e+00
sqft_living      2.079900e+03
sqft_lot         1.510697e+04
floors           1.494309e+00
waterfront       7.541757e-03
view             2.343034e-01
condition        3.409430e+00
grade            7.656873e+00
sqft_above       1.788396e+03
sqft_basement    2.915090e+02
yr_built         1.971005e+03
yr_renovated     8.440226e+01
dtype: float64

In [19]:
houses['yr_built'].mean()

1971.0051357978994

In [20]:
houses['yr_renovated'].min()

0

In [21]:
houses['yr_renovated'].max()

2015

In [22]:
houses['yr_renovated'].unique()

array([   0, 1991, 2002, 2010, 1999, 1992, 2013, 1994, 1978, 2005, 2008,
       2003, 1984, 1954, 2014, 2011, 1974, 1983, 1945, 1990, 1988, 1957,
       1977, 1981, 1995, 2000, 1998, 1970, 1989, 2004, 1986, 2009, 2007,
       1987, 1973, 2006, 1985, 2001, 1980, 1971, 1979, 1997, 1950, 1969,
       1948, 2015, 1968, 2012, 1963, 1951, 1993, 1962, 1996, 1972, 1953,
       1955, 1982, 1956, 1940, 1976, 1946, 1975, 1958, 1964, 1959, 1960,
       1967, 1965, 1934, 1944], dtype=int64)

In [23]:
houses['yr_renovated'].value_counts().head()

0       20699
2014       91
2013       37
2003       36
2000       35
Name: yr_renovated, dtype: int64

In [24]:
houses['yr_renovated'].value_counts().tail(10)

1967    2
1946    2
1971    2
1940    2
1954    1
1934    1
1959    1
1951    1
1948    1
1944    1
Name: yr_renovated, dtype: int64

In [25]:
houses['yr_built'].min()

1900

In [26]:
houses['yr_built'].max()

2015

In [27]:
houses['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

In [28]:
houses['waterfront'].value_counts()

0    21450
1      163
Name: waterfront, dtype: int64

In [29]:
houses['view'].value_counts()

0    19489
2      963
3      510
1      332
4      319
Name: view, dtype: int64

In [30]:
houses['grade'].value_counts()

7     8981
8     6068
9     2615
6     2038
10    1134
11     399
5      242
12      90
4       29
13      13
3        3
1        1
Name: grade, dtype: int64

###### Пишите выводы

В среднем:  
    
- 3-4 комнаты  
- 2+ ванных  
- 2080 футов жилой площади  
- $540088 цена  
- 3.4 оценка по 5-балльной шкале  
- 1.5 этажа  

Большинство домов не было отремонтировано совсем, 
самый богатый для ремонтов год - 2014 (91 ремонт),
в такие годы, как 1934, 1944, 1954 и др. был отремонтирован только 1 дом

Самый старый дом построен в 1900 году.
Самый новый - в 2015.
В среднем год постройки домов - 1971.

50% домов получило оценку "3",
25% домов получило оценку "4" и "5" по 5-балльной шкале.

У 163 домов есть вид на море.

У большинства домов нет вида из окна, у 963 домов вид в 2 стороны, у 319 - в 4 стороны

## Условие 3: 3 задача

### Ответьте на несколько вопросов





###### 3.1 В каком диапазоне изменяются стоимости недвижимости?

In [31]:
houses['price'].min()

75000.0

In [32]:
houses['price'].max()

7700000.0

стоимости недвижимости изменяются в диапазоне [75000, 7700000] долларов

###### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?


In [33]:
sq = houses[['sqft_living','sqft_lot']]
sq.head()

sqft_living  sqft_lot
0         1180      5650
1         2570      7242
2          770     10000
3         1960      5000
4         1680      8080

In [34]:
sq['percentage of living space'] = sq['sqft_living']/sq['sqft_lot']
sq.head()

C:\Users\innam\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


sqft_living  sqft_lot  percentage of living space
0         1180      5650                    0.208850
1         2570      7242                    0.354874
2          770     10000                    0.077000
3         1960      5000                    0.392000
4         1680      8080                    0.207921

In [35]:
sq['percentage of living space'].mean()

0.32374472542047195

жилая площадь от всей площади по всем домам в среднем занимает 32.37%

###### 3.3 Как много домов с разными этажами в данных?

In [36]:
houses['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

###### 3.4 Насколько хорошие состояния у домов в данных?

In [37]:
houses['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

In [38]:
houses['condition'].min()

1

In [39]:
houses['condition'].mean()

3.4094295100171195

In [40]:
houses['condition'].max()

5

###### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [41]:
houses['yr_built'].min()

1900

In [42]:
houses['yr_built'].max()

2015

## Условие 4: 4 задача

###### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [43]:
houses[houses['bedrooms']==2].price.mean()

401372.681884058

###### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [44]:
houses[houses['price']>600000].sqft_lot.mean()

20442.524776214832

###### 4.3 Как много домов коснулся ремонт?

In [45]:
len(houses[~(houses['yr_renovated']==0)])

914

###### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [46]:
gr_greater10 = houses[houses['grade']>10].price.mean()
gr_greater10

1678635.1175298805

In [47]:
gr_less4 = houses[houses['grade']<4].price.mean()
gr_less4

189750.0

In [48]:
gr_greater10 - gr_less4

1488885.1175298805

в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4 на 1488885.1

## Условие 5: 5 задача

###### 5.1 Выберите дом клиенту

###### Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [49]:
houses[(houses['waterfront']==True) & (houses['bathrooms']>=3) & (houses['sqft_basement']>0)]

id             date      price  bedrooms  bathrooms  \
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
1261   7636800041  20140625T000000   995000.0         3       4.50   
1315   7558700030  20150413T000000  5300000.0         6       6.00   
1361   1925069082  20150511T000000  2200000.0         5       4.25   
2626   7738500731  20140815T000000  4500000.0         5       5.50   
2864   4114601570  20141118T000000  3600000.0         3       3.25   
3914   9808700762  20140611T000000  7062500.0         5       4.50   
3934   1269200229  20140723T000000  1379900.0         3       3.25   
5760    623059016  20140717T000000  1100000.0         4       3.25   
6233   2024069008  20140619T000000  2200000.0         5       4.75   
6508   4217402115  20150421T000000  3650000.0         6       4.75   
7989   9362000040  20140623T000000  3400000.0         3       4.50   
8092   1924059029  20140617T000000  4668000.0         5       6.75   
8163   1724069079  20150319T000000  1452000.0         2       3.25   
8784   3343302110  20150306T000000  1800000.0         3       3.00   
10263   625059036  20140813T000000  2700000.0         5       4.00   
11257  3761100045  20140618T000000  3000000.0         4       4.25   
11402  3426049284  20140819T000000  2300000.0         4       3.25   
11424  5119400075  20140620T000000   950000.0         3       3.25   
11446   624069035  20141209T000000  2750000.0         4       4.00   
11844  2623039082  20150218T000000   770000.0         3       3.50   
12459  4114601580  20140724T000000  1900000.0         6       4.00   
12935  3222049044  20140612T000000   835000.0         3       3.00   
13710  2923500230  20141216T000000  2600000.0         4       4.50   
13967  3126059023  20150303T000000  3395000.0         4       3.50   
14083  3625059043  20140904T000000  3300000.0         5       4.75   
14619  7936500109  20140725T000000  2230000.0         3       3.00   
14822  3623500408  20150330T000000  2600000.0         3       3.00   
15482   624069108  20140812T000000  3200000.0         4       3.25   
15668  4030100290  20141001T000000  1680000.0         5       3.50   
16817   825059178  20140923T000000  2574000.0         4       3.75   
17767  7352200450  20150115T000000  2050000.0         4       3.25   
18199  3625059152  20141230T000000  3300000.0         3       3.25   
18455  8043700300  20140608T000000  2700000.0         4       3.25   
18556  6613000930  20140902T000000  2950000.0         4       3.25   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
300           4550     18641     1.0           1     4  ...     10   
457           1970     20978     2.0           1     3  ...      9   
1164          8010     45517     2.0           1     4  ...     12   
1261          4380     47044     2.0           1     3  ...      9   
1315          7390     24829     2.0           1     4  ...     12   
1361          4640     22703     2.0           1     4  ...      8   
2626          6640     40014     2.0           1     4  ...     12   
2864          5020     12431     2.0           1     4  ...     10   
3914         10040     37325     2.0           1     2  ...     11   
3934          3786     38038     1.0           1     4  ...      9   
5760          3190     11774     2.0           1     4  ...      8   
6233          5990     10450     2.0           1     4  ...     11   
6508          5480     19401     1.5           1     4  ...     11   
7989          5230     17826     2.0           1     4  ...     10   
8092          9640     1

###### 5.2 Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [56]:
df = houses[((houses['waterfront']==True) | (houses['view']==4)) & (houses['condition']==5) & (houses['yr_built']>=1980)]
df

id             date      price  bedrooms  bathrooms  \
11257  3761100045  20140618T000000  3000000.0         4       4.25   
14682  7278100515  20140821T000000  1295000.0         2       2.50   
17312  3585901025  20140613T000000  1735000.0         3       2.50   
18991  7856400300  20140702T000000  1411600.0         2       2.50   
18992  7856400300  20150322T000000  1505000.0         2       2.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
11257         4850     12445     2.0           1     4  ...     10   
14682         2910     19449     2.0           1     4  ...      9   
17312         4310     32093     1.5           0     4  ...     10   
18991         3180      9400     2.0           0     4  ...     10   
18992         3180      9400     2.0           0     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
11257      3850.0           1000      1989             0    98034  47.7011   
14682      1940.0            970      1985             0    98177  47.7729   
17312      2510.0           1800      1982             0    98177  47.7624   
18991      2610.0            570      1985             0    98006  47.5617   
18992      2610.0            570      1985             0    98006  47.5617   

          long  sqft_living15  sqft_lot15  
11257 -122.244           3350       12210  
14682 -122.393           2540       23598  
17312 -122.380           3810       26400  
18991 -122.158           3760        9450  
18992 -122.158           3760        9450  

[5 rows x 21 columns]

In [57]:
df.price.min()

1295000.0

In [58]:
df.price.max()

3000000.0

###### 5.3 Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [59]:
df2 = houses[(houses['sqft_basement']==0) & (houses['floors']==2) & (houses['price']<150000)]
df2

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2589           980      5000     2.0           0     0  ...      7   
10105         1230      1561     2.0           0     0  ...      7   
15610         1270      8991     2.0           0     0  ...      7   
18332         1200      7000     2.0           0     0  ...      7   
18939          980      2490     2.0           0     0  ...      6   
18988         1200      2002     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2589        980.0              0      1922          2003    98014  47.7076   
10105      1230.0              0      1983             0    98042  47.3768   
15610      1270.0              0      1981             0    98042  47.3563   
18332      1200.0              0      1908             0    98116  47.5883   
18939       980.0              0      1969             0    98092  47.3170   
18988      1200.0              0      1966             0    98055  47.4659   

          long  sqft_living15  sqft_lot15  
2589  -121.359           1040        5000  
10105 -122.149           1660        2243  
15610 -122.149           1270        8993  
18332 -122.384           3290        6000  
18939 -122.182            980        3154  
18988 -122.189           1270        1848  

[6 rows x 21 columns]

In [60]:
df2.condition.mean()

2.8333333333333335